### PR TITLE
Fix: in UI5, classes can act as namespaces

### LINKED ISSUE #48: dts-generator ignores child namespaces/enums etc. of classes
In UI5, classes can have namespaces, classes, enums, interfaces or typedefs as static children.
Examples:
* [sap.ui.core.Popup.Dock](https://openui5nightly.hana.ondemand.com/api/sap.ui.core.Popup.Dock) (namespace, rather should be an enum - never mind)
* [sap.ui.core.View.Preprocessor](https://openui5nightly.hana.ondemand.com/api/sap.ui.core.mvc.View.Preprocessor) (interface)
* [sap.ui.core.XMLView.PreprocessorType](https://openui5nightly.hana.ondemand.com/api/sap.ui.core.mvc.XMLView.PreprocessorType) (enum)
* [sap.ui.core.UIComponent.RoutingMetadata](https://openui5nightly.hana.ondemand.com/api/sap.ui.core.UIComponent.RoutingMetadata) (type definition)

The dts-generator currently ignores such children. It assumes, that only namespaces can contain namespaces, classes etc. When such types are referenced in signatures, they are reported as unresolved dotted type names and replaced with type `any`. 

According to the following [stackoverflow question](https://stackoverflow.com/questions/29844959/enum-inside-class-typescript-definition-file),  it should be possible to represent the class as both, a `class` and a `namespace` in such cases. All nested namespaces, classes, enums etc. should go into the `namespace` declaration, all instance methods and fields as well as simple static members (functions, fields) should go into the `class` definition.

I tried the following in VS code (patching the sap.ui.core.d.ts) and it worked for me
```js
  namespace Popup {
    enum Dock { ... }
  }
  class Popup {
    open()
    close()
  }
```

### PR DESCRIPTION
**Pull request by codeworrior**
In UI5, classes can act as namespaces, too. They can contain other namespaces, classes or enums as static members.

When converting the UI5 api.json into a tree structure, children of classes therefore have to be searched and added, too.

In the resulting d.ts file, classes can appear twice after this change:
 - as a `namespace`, containing nested namespaces, classes,  interfaces and enums
 - and as a `class`, containing all other kinds of members

Fixes #48. At least partially. Support for typedefs is currently missing in the dts-generator, this is not fixed by this PR.

### CODE REVIEW

File: packages/dts-generator/lib/phases/json-to-ast.js [hunk 1]
```diff
@@ -11,51 +11,75 @@ function jsonToAst(jsonObj) {
 /**
  * @param symbols
  * @param {string} nsFQN
+ * @param {boolean} [isClass=false]
  * @returns {Namespace}
  */
-function buildNamespace(symbols, nsFQN) {
+function buildNamespace(symbols, nsFQN, isClass) {
```
Comment (bd82) [inline packages/dts-generator/lib/phases/json-to-ast.js]: Should we have two functions instead of a boolean flag
with multiple conditionals?

**Comment by bd82** [comment]
Build fails with strange:
> lerna ERR! MaxBufferError: stdout maxBuffer exceeded

**Comment by bd82** [comment]
also as with #50 

If you also execute the gen:dts script.

    /packages/ts-types/package.json@master#L20

We could see/review how this affects the generated output as part of the PR.

File: packages/dts-generator/lib/phases/json-to-ast.js (hunk 1 above)
Comment (codeworrior) [reply packages/dts-generator/lib/phases/json-to-ast.js]: Done

**Comment by codeworrior** [comment]
The pre-commit hook didn't work for me. I've run now gen:dts manually.

Works as expected, but in 1.60, the `Popup.Dock` is modeled as a namespace and this doesn't work out in the generated d.ts file. Should I manually map it to `any` again in your directives? In the actual master of UI5, I'll change `Popup.Dock` to a proper enum.

**Comment by bd82** [comment]
> Works as expected, but in 1.60, the Popup.Dock is modeled as a namespace and this doesn't work out in the generated d.ts file. Should I manually map it to any again in your directives? In the actual master of UI5, I'll change Popup.Dock to a proper enum.

Sounds like a plan 👍

**Comment by tmundt** [comment]
This PR is still open. What is this failure w/ CircleCI?

**Comment by bd82** [comment]
The generated artifacts now include compilation errors after this fix...

**Comment by bd82** [comment]
closing this for now.

Feel free to re-open if you have time to re-tackle this and rebase it.

**Pull request closed without merging**